File: app/src/main/java/com/example/win/easy/view/main/LoginFragment.java
package com.example.win.easy.view.main;

import android.os.Bundle;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.MutableLiveData;
import androidx.navigation.Navigation;

import com.example.win.easy.R;
import com.example.win.easy.enumeration.LoginType;
import com.example.win.easy.web.callback.OnReadyFunc;
import com.example.win.easy.web.service.LoginService;
import com.qmuiteam.qmui.alpha.QMUIAlphaButton;
import com.qmuiteam.qmui.widget.QMUITopBar;
import com.qmuiteam.qmui.widget.dialog.QMUITipDialog;

import butterknife.BindInt;
import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;

import static com.example.win.easy.enumeration.LoginType.Email;
import static com.example.win.easy.enumeration.LoginType.Phone;

public class LoginFragment extends Fragment {

    @BindView(R.id.loginTopBar) QMUITopBar topBar;
    @BindView(R.id.accountEditText) EditText accountEditText;
    @BindView(R.id.passwordEditText) EditText passwordEditText;
    @BindView(R.id.loginButton) QMUIAlphaButton loginButton;
    @BindView(R.id.switchLoginTypeButton)QMUIAlphaButton switchLoginTypeButton;
    @BindString(R.string.phoneAccountHint) String phoneAccountHint;
    @BindString(R.string.emailAccountHint) String emailAccountHint;
    @BindString(R.string.switchToPhoneLoginHint) String switchToPhoneLoginHint;
    @BindString(R.string.switchToEmailLoginHint) String switchToEmailLoginHint;
    @BindString(R.string.loginingTipWord)String loginingTipWord;
    @BindString(R.string.successfulLoginTipWord) String successfulLoginTipWord;
    @BindString(R.string.unsuccessfulLoginTipWord) String unsuccessfulLoginTipWord;
    @BindString(R.string.loginFragmentTopBarTitle) String topBarTitle;
    @BindInt(R.integer.temporaryTipShowTimeInMillis)int tipShowTimeInMillis;

    private LoginService loginService;
    private MutableLiveData<LoginType> loginTypeLiveData=new MutableLiveData<>();
    private QMUITipDialog loginingDialog;
    private LoginType defaultLoginType;

    public LoginFragment(LoginService loginService,LoginType defaultLoginType){
        this.loginService=loginService;
        this.defaultLoginType=defaultLoginType;
    }

    /**
     *
     */
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View thisView=inflater.inflate(R.layout.fragment_login,container,false);
        ButterKnife.bind(this,thisView);
        setTopBar(topBarTitle);
        setUpLoginButton();
        setUpSwitchLoginTypeButton();
        setUpLoginType();
        return thisView;
    }

    private void setUpLoginType() {
        loginTypeLiveData.setValue(defaultLoginType);
        loginTypeLiveData.observe(this,this::onLoginTypeChanged);
    }

    private void onLoginTypeChanged(LoginType loginType){
        if (loginType==Phone){
            accountEditText.setHint(phoneAccountHint);
            accountEditText.setInputType(InputType.TYPE_CLASS_PHONE);
            switchLoginTypeButton.setText(switchToEmailLoginHint);
        }else {
            accountEditText.setHint(emailAccountHint);
            accountEditText.setInputType(InputType.TYPE_TEXT_VARIATION_WEB_EMAIL_ADDRESS);
            switchLoginTypeButton.setText(switchToPhoneLoginHint);
        }
    }

    private void setTopBar(String title){
        topBar.setTitle(title);
    }

    private void setUpLoginButton(){
        loginButton.setOnClickListener(this::login);
    }

    private void login(View v){
        showLoginingDialog();
        login(getAccount(),getPassword(), loginTypeLiveData.getValue());
    }

    private void showLoginingDialog(){
        loginingDialog=buildDialog(loginingTipWord,QMUITipDialog.Builder.ICON_TYPE_LOADING);
        loginingDialog.show();
    }

    private QMUITipDialog buildDialog(String tipWord,int iconType){
        return new QMUITipDialog.Builder(getContext())
                .setIconType(iconType)
                .setTipWord(tipWord)
                .create();
    }

    private void login(String account,String password,LoginType loginType){
        switch (loginType){
            case Phone: loginService.loginByPhone(account,password, loginFinishedCallback());
                break;
            case Email: loginService.loginByEmail(account,password, loginFinishedCallback());
                break;
            default:
        }
    }

    private OnReadyFunc<Boolean> loginFinishedCallback(){
        return successful->{
            if (successful) onSuccessfulLogin();
            else onUnsuccessfulLogin();
        };
    }

    private String getAccount(){
        return accountEditText.getText().toString();
    }

    private String getPassword(){
        return passwordEditText.getText().toString();
    }

    private void onSuccessfulLogin(){
        dismissLoginingDialog();
        temporarilyShowSuccessfulLoginTip(successfulLoginTipWord,tipShowTimeInMillis);
        navigateToMainFragment();
    }
    private void onUnsuccessfulLogin(){
        dismissLoginingDialog();
        temporarilyShowUnsuccessfulLoginTip(unsuccessfulLoginTipWord,tipShowTimeInMillis);
    }

    private void dismissLoginingDialog(){
        loginingDialog.dismiss();
    }

    private void temporarilyShowSuccessfulLoginTip(String tipWord, int showTimeInMillis){
        temporarilyShowTipDialog(tipWord,QMUITipDialog.Builder.ICON_TYPE_SUCCESS,showTimeInMillis);
    }

    private void temporarilyShowUnsuccessfulLoginTip(String tipWord, int showTimeInMillis){
        temporarilyShowTipDialog(tipWord,QMUITipDialog.Builder.ICON_TYPE_FAIL,showTimeInMillis);
    }

    private void temporarilyShowTipDialog(String tipWord,int iconType,  int showTimeInMillis){
        QMUITipDialog tipDialog=buildDialog(tipWord,iconType);
        tipDialog.show();
        dismissAfter(tipDialog,showTimeInMillis);
    }

    private void dismissAfter(QMUITipDialog dialog, int showTimeInMillis){
        getView().postDelayed(dialog::dismiss,showTimeInMillis);
    }

    private void navigateToMainFragment(){
        Navigation.findNavController(getView()).navigate(LoginFragmentDirections.actionLoginFragmentToMainActivityFragment());
    }

    private void setUpSwitchLoginTypeButton() {
        switchLoginTypeButton.setOnClickListener(v -> {
            LoginType currentLoginType=loginTypeLiveData.getValue();
            if (currentLoginType==Phone)
                loginTypeLiveData.setValue(Email);
            else
                loginTypeLiveData.setValue(Phone);
        });
    }


}

File: app/src/main/java/com/example/win/easy/view/main/LoginFragment.kt
package com.example.win.easy.view.main

import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ProgressBar
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.fragment.app.Fragment
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.navigation.NavController
import androidx.navigation.Navigation
import butterknife.BindString
import butterknife.BindView
import butterknife.ButterKnife
import com.example.win.easy.R
import com.example.win.easy.enumeration.LoginType
import com.example.win.easy.network.LoginService
import com.google.android.material.snackbar.Snackbar
import com.qmuiteam.qmui.alpha.QMUIAlphaButton
import com.qmuiteam.qmui.widget.QMUITopBar
import kotlinx.coroutines.*

class LoginFragment(private val loginService: LoginService, private val defaultLoginType: LoginType)
    : Fragment(),
        CoroutineScope by CoroutineScope(Dispatchers.Main){
    @BindView(R.id.login_fragment_root_layout) lateinit var rootLayout: ConstraintLayout
    @BindView(R.id.loginTopBar) lateinit var topBar: QMUITopBar
    @BindView(R.id.accountEditText) lateinit var accountEditText: EditText
    @BindView(R.id.passwordEditText) lateinit var passwordEditText: EditText
    @BindView(R.id.loginProgressBar) lateinit var spinner: ProgressBar
    @BindView(R.id.loginButton) lateinit var loginButton: QMUIAlphaButton
    @BindView(R.id.switchLoginTypeButton) lateinit var switchLoginTypeButton: QMUIAlphaButton
    @BindString(R.string.phoneAccountHint) lateinit var phoneAccountHint: String
    @BindString(R.string.emailAccountHint) lateinit var emailAccountHint: String
    @BindString(R.string.switchToPhoneLoginHint) lateinit var switchToPhoneLoginHint: String
    @BindString(R.string.switchToEmailLoginHint) lateinit var switchToEmailLoginHint: String
    @BindString(R.string.loginingTipWord) lateinit var loginingTipWord: String
    @BindString(R.string.successfulLoginTipWord) lateinit var successfulLoginTipWord: String
    @BindString(R.string.unsuccessfulLoginTipWord) lateinit var unsuccessfulLoginTipWord: String
    @BindString(R.string.loginFragmentTopBarTitle) lateinit var topBarTitle: String
    private val loginTypeLiveData = MutableLiveData<LoginType>()
    private val account: String
        get() = accountEditText.text.toString()

    private val password: String
        get() = passwordEditText.text.toString()

    /**
     *
     */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.fragment_login, container, false).apply {
                ButterKnife.bind(this@LoginFragment, this)
                topBar.setTitle(topBarTitle)
                loginButton.setOnClickListener { login(account, password, loginTypeLiveData.value!!) }
                switchLoginTypeButton.setOnClickListener {
                    loginTypeLiveData.value = if (loginTypeLiveData.value == LoginType.Email) LoginType.Phone else LoginType.Email
                }
                loginTypeLiveData.run {
                    value=defaultLoginType
                    observe(this@LoginFragment, Observer { onLoginTypeChanged(it) })
                }
            }

    private fun onLoginTypeChanged(loginType: LoginType): Unit =
        if (loginType == LoginType.Phone) {
            accountEditText.hint = phoneAccountHint
            accountEditText.inputType = InputType.TYPE_CLASS_PHONE
            switchLoginTypeButton.text = switchToEmailLoginHint
        } else {
            accountEditText.hint = emailAccountHint
            accountEditText.inputType = InputType.TYPE_TEXT_VARIATION_WEB_EMAIL_ADDRESS
            switchLoginTypeButton.text = switchToPhoneLoginHint
        }

    private fun login(account: String, password: String, loginType: LoginType) {
        launch {
            try {
                spinner.visibility = ProgressBar.VISIBLE
                withTimeout(5000){
                    loginService.login(account, password,loginType)
                }
                Snackbar.make(rootLayout, successfulLoginTipWord, Snackbar.LENGTH_SHORT).show()
                Navigation.findNavController(view!!).navigateUp()
            } catch (t: Throwable) {
                Snackbar.make(rootLayout, "登录失败：" + t.message, Snackbar.LENGTH_SHORT).show()
            } finally {
                spinner.visibility = ProgressBar.GONE
            }
        }
    }
}